schema/gen/go/generate.go

```diff
@@ -1,14 +1,133 @@
+	"fmt"
 	"io"
 	"os"
 	"path/filepath"
+	"sort"
 
 	"github.com/ipld/go-ipld-prime/schema"
 )
 
+// Generate takes a typesystem and the adjunct config for codegen,
+// and emits generated code in the given path with the given package name.
+//
+// All of the files produced will match the pattern "ipldsch.*.gen.go".
 func Generate(pth string, pkgName string, ts schema.TypeSystem, adjCfg *AdjunctCfg) {
+	// Emit fixed bits.
+	withFile(filepath.Join(pth, "ipldsch.minima.gen.go"), func(f io.Writer) {
+		EmitInternalEnums(pkgName, f)
+	})
+
+	// Local helper function for applying generation logic to each type.
+	//  We will end up doing this more than once because in this layout, more than one file contains part of the story for each type.
+	applyToEachType := func(fn func(tg TypeGenerator, w io.Writer), f io.Writer) {
+		// Sort the type names so we have a determinisic order; this affects output consistency.
+		//  Any stable order would do, but we don't presently have one, so a sort is necessary.
+		types := ts.GetTypes()
+		keys := make(sortableTypeNames, 0, len(types))
+		for tn := range types {
+			keys = append(keys, tn)
+		}
+		sort.Sort(keys)
+		for _, tn := range keys {
+			switch t2 := types[tn].(type) {
+			case *schema.TypeBool:
+				fn(NewBoolReprBoolGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeInt:
+				fn(NewIntReprIntGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeFloat:
+				fn(NewFloatReprFloatGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeString:
+				fn(NewStringReprStringGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeBytes:
+				fn(NewBytesReprBytesGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeLink:
+				fn(NewLinkReprLinkGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeStruct:
+				switch t2.RepresentationStrategy().(type) {
+				case schema.StructRepresentation_Map:
+					fn(NewStructReprMapGenerator(pkgName, t2, adjCfg), f)
+				case schema.StructRepresentation_Tuple:
+					fn(NewStructReprTupleGenerator(pkgName, t2, adjCfg), f)
+				case schema.StructRepresentation_Stringjoin:
+					fn(NewStructReprStringjoinGenerator(pkgName, t2, adjCfg), f)
+				default:
+					panic("unrecognized struct representation strategy")
+				}
+			case *schema.TypeMap:
+				fn(NewMapReprMapGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeList:
+				fn(NewListReprListGenerator(pkgName, t2, adjCfg), f)
+			case *schema.TypeUnion:
+				switch t2.RepresentationStrategy().(type) {
+				case schema.UnionRepresentation_Keyed:
+					fn(NewUnionReprKeyedGenerator(pkgName, t2, adjCfg), f)
+				case schema.UnionRepresentation_Kinded:
+					fn(NewUnionReprKindedGenerator(pkgName, t2, adjCfg), f)
+				default:
+					panic("unrecognized union representation strategy")
+				}
+			default:
+				panic("add more type switches here :)")
+			}
+		}
+	}
+
+	// Emit a file with the type table, and the golang type defns for each type.
+	withFile(filepath.Join(pth, "ipldsch.types.gen.go"), func(f io.Writer) {
```

Comment:
not sure I like "ipldsch", it reads like some germanic gibberish to me :) 

how about just ipldschema_types.go? it's self-describing, not too long, and honestly I don't think we need the `gen.go` suffix if we already have an `ipldschema_` prefix. underscores before the extension dot are also generally more common.

Reply:
I thought (but maybe I imagined) I'd seen a convention ".gen" as a prefix to the file extension suffix as a semi-universal hint that a file is generated.  I wish some convention of that existed, anyway.  But maybe that was wishful thinking?

Reply:
The convention is the "Code generated by..." comment, and nothing else, really. Take stringer, which generates `yourtype_string.go`.

I think a prefix like `ipldschema_` or `ipldsch_` is already extremely clear, to be honest. Adding `.gen` just adds verbosity for a convention that doesn't exist :)